plugins/action_generate_review_assignment_records/src/generateReviewAssignments.ts

```diff
@@ -146,17 +146,47 @@ const generateNextReviewAssignments = async ({
   console.log('Next level reviewers', nextLevelReviewers)
   const reviewAssignments: ReviewAssignmentObject = {}
 
+  // Check if other reviewAssignmet is already assigned to create new ones LOCKED
+  const existingReviewAssignments: ExistingReviewAssignment[] =
+    await db.getExistingReviewAssignments(applicationId, stageNumber, nextReviewLevel)
+
+  const getNewOrExistingAssignmentStatus = (
+    userId: number,
+    canMakeFinalDecision: boolean,
+    canSelfAssign: boolean
+  ) => {
+    // Temporarly final decision shouldn't be locked if there are other reviewAssignemt assigned
+    if (canMakeFinalDecision) return { status: ReviewAssignmentStatus.Assigned, isLocked: false }
+
+    // Check if existing review assignment
+    const existingAssignment = existingReviewAssignments.find(
+      (reviewAssignment) => reviewAssignment.userId === userId
+    )
+    // If user already has reviewAssignment return same status and isLocked flag
+    if (existingAssignment)
+      return { status: existingAssignment.status, isLocked: existingAssignment.isLocked }
+
+    // Non-existing review assignments
+    const isLocked = existingReviewAssignments.some(
+      ({ status }) => status == ReviewAssignmentStatus.Assigned
+    )
```

Comment:
It seems inefficient to do the same search through the same array for every reviewer. Can this value just be deduced once for all?

Maybe just below `const existingReviewAssignments = ...`, but outside the loop:

```js
const anyAlreadyAssigned = existingReviewAssignments.some(
      ({ status }) => status == ReviewAssignmentStatus.Assigned
    )
```

And then just refer to `anyAlreadyAssigned` inside the loop?
```js
if (canSelfAssign || nextReviewLevel > 1)
  return { status: ReviewAssignmentStatus.AvailableForSelfAssignment, isLocked: anyAlreadyAssigned }
return {
  status: ReviewAssignmentStatus.Available,
  isLocked: anyAlreadyAssigned,
}
```

I think that would be easier to understand, too, personally.